### ISSUE TITLE
Cluster Explorer caused denial of service in downstream cluster

### ISSUE THREAD

**Issue by samjustus** [rancher/rancher#37627]
SURE-3208

**Rancher Server Setup**
- Rancher version: 2.6.4

**Describe the bug**
Since upgrading to 2.6.4 after https://github.com/rancher/rancher/issues/36359 was implemented, (which we thought would fix the issue) they have continued to see several spikes.

April 4 9:00pm ~55k events per 15m
April 5 2:30am ~15k events per 15m
April 6 3:30am ~35k events per 15m
April 7 1am-7am ~15k events per 15m
April 7 9am ~40k events per 15m
April 10 3:00pm ~ 40k events per 15m
   
The requests always come from the cattle-cluster-agent pod
As best as we can tell, it is only from users with an open browsing session. But we've done packet traces (though this was back in 2.5) and the browser is not sending any data to Rancher at all during the spike in kube-apiserver calls.  The kube-apiserver calls are executed but not passed along to the browser.  As soon as the browser window is closed, the kube-apiserver traffic subsides immediately.

The requests appear to be legitimate, in the sense that they appear to be API calls that the Rancher UI would be making.  They're mostly enumeration calls  – giant floods of GET calls for all the workloads in every namespace that the user has access to.  And they repeat over and over and over.

For example, taking one spike for a single user, I did an extract from the URI:

rex field=URI "/apis/apps/v1/namespaces/[^/]+/(?<type>.*)?"
--
And the result was that this one user did a GET matching that pattern:
replicasets | 116677
deployments | 110876
NONE | 20619
daemonsets | 2270
statefulsets | 2270
That means that during a ~1 hour period, cattle-cluster-agent did 116,667 "GET" calls to retrieve a listing of replicasets, 110,867 "GET" calls to retrieve a listing of deployments, etc.  That's an INSANE number of calls for a single user.  So no, they're not "genuine" in the sense that I don't think the user (or the user's browser) is legitimately asking for this data.
It is pretty clear to me that cattle-cluster-agent gets stuck into some kind of weird state where it's trying to send the browser all the latest data so it can render the workloads UI, but instead of just getting the data once and stopping, it fetches the data (which on its own is a pretty big set of API calls) and then instead of waiting to be told to refresh, it just goes into a tight loop and makes the same set of requests over and over and over again.
My guess is that this has something to do with the websocket connection between the browser and Rancher, because we don't see evidence of these API calls between the browser and Rancher.  The kube-apiserver delivers responses to all these calls then cattle-cluster-agent just seems to throw them away.  Or maybe it sends them to rancher and then rancher throws them away.  But they never make it to the user's browser.

**To Reproduce**
Open Cluster Explorer for one of the downstream clusters in a browser
Monitor kube-apiserver logs.

**Result**
The cattle-cluster-agent seems to be spamming the kube-apiserver.

**Expected Result**
Viewing the downstream cluster does not impact performance.

**Comment by cbron** [comment]
This is on hold while ^ 5997 is investigated.

**Comment by anupama2501** [comment]
Reproduced on v2.6.4

1. Create a downstream cluster.
2. Turn on debug logging for the agent by setting CATTLE_DEBUG to true in the environment variable configuration for the cluster from advanced options >> add env variable >> key=CATTLE_DEBUG, value=true
3. Set the env variable `CATTLE_WATCH_TIMEOUT_SECONDS` to `300`
4. Turn on verbose logging for the kubernetes API server by editing the YAML, and under rancher_kubernetes_engine_config->services-> kube-api add a section
```
kube-api:
  extra_args:
    v: '3'
```
5. When the cluster is active, navigate to “Deployments” under “Workloads”.
6. Created 100 deployments, and navigate to workloads tab.
7. Open the Network tab >> Wait for 5 minutes >> check the console logs:

<img width="400" alt="socke" src="https://user-images.githubusercontent.com/38144301/178404160-535cb2fe-942c-448c-b5c0-af8329f62da7.png">

**Comment by anupama2501** [comment]
Re-opened as an upgrade from 2.6.5 >> 2.6-head 3506e43 on a fresh install v2.6-head `3506e43`
Same steps as the above comment. Do not see the error received through websocket

<img width="300" alt="socket3" src="https://user-images.githubusercontent.com/38144301/178765489-03605482-3b53-44b8-b733-248e5e11c382.png">

See some watch errors logs in rancher-agent logs:
```
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for node"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for apps.statefulset"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for configmap"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for pod"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for batch.job"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for apps.deployment"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for apps.daemonset"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for event"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for apps.replicaset"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for apps.deployment"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for service"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for batch.cronjob"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: context canceled\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for persistentvolumeclaim"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for secret"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding"
time="2022-07-12T23:10:13Z" level=debug msg="closing watcher for replicationcontroller"
time="2022-07-12T23:10:13Z" level=debug msg="event watch error: an error on the server (\"unable to decode an event from the watch stream: http2: response body closed\") has prevented the request from succeeding"
```

**Comment by cmurphy** [comment]
@anupama2501 I looked on your cluster and I do not see the watch rapidly opening and closing, which is the bug. On your setup I notice that you have 60 deployments running. From the reproduction steps in https://github.com/rancher/dashboard/issues/5997 notice that you have to create AND delete them before starting to monitor the events, because the resource version for the individual deployments needs to be significantly different from the resource version for the deployment list in order to trigger the "too old" error.

The "unable to decode an event from the watch stream: context canceled" message in the agent logs is not from the dashboard itself opening and closing the watch, it is from the dashboard being refreshed and is normal.

**Comment by samjustus** [comment]
@cmurphy @anupama2501 UI never got around to completing their end of things. Are we still able to QA this then? Should we? Would we be able to reliably close the jira issue without the dashboard issue being resolved?

https://github.com/rancher/dashboard/issues/5997

**Comment by anupama2501** [comment]
Verified on v2.6.7-rc2
Rancher server HA helm install
1. Created a downstream rke1 cluster with 1 worker , 1 etcd, 1 cp
2. While creating the cluster, in the env variables added CATTLE_DEBUG=true and CATTLE_WATCH_TIMEOUT_SECONDS=600
3. Created a 30 workloads from deployments page
4. Created 2 stateful set, installed monitoring v2 app
5. Deleted a few workloads
6. Navigated to workloads tab and waited for a few minutes, noticed the watch close and open rapidly and see the error :
```
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for batch. cronjob"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for apps. statefulset"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for replicationcontroller"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for node"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for namespace"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for batch. job"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for apps. daemonset"
 Sat, Jul 16 2022 11:53:13 am time="2022-07-16T18:53:13Z" level=debug msg="opening watcher for ui.cattle.io. navlink"
 Sat, Jul 16 2022 11:53:15 am time="2022-07-16T18:53:15Z" level=debug msg="Wrote ping
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg= opening watcher for apps. replicaset"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="opening watcher for apps. deployment"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="event watch error: too old resource version: 195821 (197732)"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="closing watcher for apps.deployment"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="event watch error:   too old resource version: 195820  (197621)"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="closing watcher for apps.replicaset"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="opening watcher for pod"
 Sat, Jul 16 2022 11:53:16 am time="2022-07-16T18:53:16Z" level=debug msg="event watch error: too old resource version: 197141 (198290)"
```

<img width="400" alt="watcherror" src="https://user-images.githubusercontent.com/38144301/179368748-38377359-bbac-418d-83a7-6bda116c5eff.png">

**Comment by cmurphy** [comment]
## Release note template

**Synopsis of issue**: if the Cluster Explorer dashboard is left open to one page for a downstream cluster for more than 30 minutes, it starts rapidly opening and closing watch requests perpetually, eventually overloading the downstream kubernetes API server.
**Resolved (or not)**: yes
**Workaround**: (for <=v2.6.6) refresh or close the browser before 30 minutes
**Versions**: v2.6.0-v2.6.6
